### PR TITLE
Integrate SendGrid with HttpClientFactory

### LINKED ISSUE #880: Dependency Injection in Startup.cs
Hi, sorry for not following the template, but regarding issues #847  and #521, I was wondering if there is any progress?

I would like to inject a SendGridClient through DI to a controller, instead of instantiating a new client every request.

I can't rely on `services.AddHttpClient<ISendGridClient, SendGridClient>();` because I need to add the `Key` that is typically called in the SendGridClient constructor.

Is this workflow being looked in to?

Thanks

### PR DESCRIPTION
**Pull request by akunzai**
# Fixes #880  

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the development branch.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added in line documentation to the code I modified

### Short description of what this PR does:
- Introduce new `SendGrid.Extensions.DependencyInjection` package for HttpClientFactory integrations

### CODE REVIEW

File: tests/SendGrid.Extensions.DependencyInjection.Tests/ServiceCollectionExtensionsTests.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        [Fact]
+        public void TestAddSendGridReturnHttpClientBuilder()
+        {
+            // Arrange
+            var collection = new ServiceCollection();
+
+            // Act
+            var builder = collection.AddSendGrid(options =>
+                options.ApiKey = Environment.GetEnvironmentVariable("SENDGRID_API_KEY") ?? "SENDGRID_API_KEY");
```
Comment (aevitas) [inline tests/SendGrid.Extensions.DependencyInjection.Tests/ServiceCollectionExtensionsTests.cs]: Wouldn't it be more efficient to forego the `Environment.GetEnvironmentVariable` call entirely, since the value possibly retrieved from it isn't being used anyway?

File: src/SendGrid.Extensions.DependencyInjection/InjectableSendGridClient.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+// </copyright>
+
+using System.Net.Http;
+using Microsoft.Extensions.Options;
+
+namespace SendGrid.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// A wrapped SendGridClient with single constructor to inject managed <see cref="HttpClient"/> by the <see cref="IHttpClientFactory"/>
```
Comment (aevitas) [inline src/SendGrid.Extensions.DependencyInjection/InjectableSendGridClient.cs]: I personally think the wording here could be a little more verbose. Something along the lines of `[...] to inject an <see cref="HttpClient"/> whose lifetime is managed externally, e.g. by an IOC container.`

File: tests/SendGrid.Extensions.DependencyInjection.Tests/ServiceCollectionExtensionsTests.cs (hunk 1 above)
Comment (akunzai) [reply tests/SendGrid.Extensions.DependencyInjection.Tests/ServiceCollectionExtensionsTests.cs]: Thanks 👌

File: src/SendGrid.Extensions.DependencyInjection/InjectableSendGridClient.cs (hunk 2 above)
Comment (akunzai) [reply src/SendGrid.Extensions.DependencyInjection/InjectableSendGridClient.cs]: Thanks, I personally prefer DI container.

**Comment by aevitas** [comment]
LGTM 👍

**Comment by akunzai** [comment]
After some investigations on Travis CI,

this PR include some building optimizations.

the building/testing time is reduced from 10 minutes to 3 minutes now.

**Comment by akunzai** [comment]
Hello @aevitas @thinkingserious,

Please let me know if any other suggestions.

File: ExampleNet45Project/Program.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            Subject = "Sending with Twilio SendGrid is Fun",
+            MailSettings = new MailSettings
+            {
+                SandboxMode = new SandboxMode
+                {
+                    Enable = configuration.GetValue("SendGrid:SandboxMode",true)
+                }
+            }
+        };
+        msg.AddContent(MimeType.Text, "and easy to do anywhere, even with C#");
```
Comment (aevitas) [inline ExampleNet45Project/Program.cs]: Har har :)

File: README.md [hunk 4]
```diff
@@ -143,34 +134,26 @@ After executing the above code, `response.StatusCode` should be `202` and you sh
...
 
 ```csharp
-using SendGrid;
-using SendGrid.Helpers.Mail;
 using System;
 using System.Threading.Tasks;
+using SendGrid;
+using SendGrid.Helpers.Mail;
 
-namespace Example
```
Comment (aevitas) [inline README.md]: In my opinion, these should be prefaced by accessors, especially if we're including `using` statements in the examples. Every class should be namespaced, even though it's kind of redundant in this example. `Main()` should be marked `private` and `class Program` should be marked `internal class Program`, for the sake of consistency.

**Comment by aevitas** [comment]
@akunzai Great job on the optimizations and upgrading the projects from netcore 1.0. I don't think anyone can be assumed to be using those runtimes at this point anymore.

I've just left some very minor comments on the readme concerning accessors. I think it would make sense to keep the code uniform in terms of accessors, but it's up to you.

File: README.md (hunk 4 above)
Comment (akunzai) [reply README.md]: IMHO, the sample code should keep it as simple as possible.

File: ExampleNet45Project/Program.cs (hunk 3 above)
Comment (akunzai) [reply ExampleNet45Project/Program.cs]: Can you please clearly point out the problem?

**Comment by IanKemp** [comment]
Sooo... is this likely to get merged anytime soon, since everyone appears to be happy with it?

**Comment by aevitas** [comment]
@IanKemp Depends on when @thinkingserious gets around to it.

File: src/SendGrid.Extensions.DependencyInjection/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+        /// <summary>
+        /// Adds the <see cref="System.Net.Http.IHttpClientFactory"/> with <see cref="ISendGridClient"/> and related services to the <see cref="IServiceCollection"/>
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="configureOptions">A delegate that is used to configure a <see cref="SendGridClientOptions"/>.</param>
+        /// <returns>An <see cref="T:Microsoft.Extensions.DependencyInjection.IHttpClientBuilder" /> that can be used to configure the client.</returns>
+        public static IHttpClientBuilder AddSendGrid(this IServiceCollection services, Action<SendGridClientOptions> configureOptions)
+        {
+            return services.AddSendGrid((resolver, options) => configureOptions(options));
```
Comment (IanKemp) [inline src/SendGrid.Extensions.DependencyInjection/ServiceCollectionExtensions.cs]: Use discard?

```
return services.AddSendGrid((_, options) => configureOptions(options));
```

File: src/SendGrid.Extensions.DependencyInjection/InjectableSendGridClient.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+using System.Net.Http;
+using Microsoft.Extensions.Options;
+
+namespace SendGrid.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// A wrapped SendGridClient with single constructor to inject an <see cref="HttpClient"/> whose lifetime is managed externally, e.g. by an DI container.
+    /// </summary>
+    public class InjectableSendGridClient : SendGridClient
```
Comment (IanKemp) [inline src/SendGrid.Extensions.DependencyInjection/InjectableSendGridClient.cs]: Should this be public, or internal?

**Comment by akunzai** [comment]
latest commit fixes:

- Floating StyleCop.Analyzers version
- Add StyleCop.Analyzers to SendGrid.Extensions.DependencyInjection
- Show StyleCop.Analyzers warnings in travis

**Comment by ryanjerskine** [comment]
Any updates on this? Seems like a valuable PR that will prevent me from rewriting much of the DI myself.

**Comment by akunzai** [comment]
Hi @thinkingserious,

Conflicts solved for #969.

Automated deploy of the `SendGrid.Extensions.DependencyInjection.*.nupkg` was ignored.

Please let me know if you prefer automated deploy the `SendGrid.Extensions.DependencyInjection.*.nupkg` too.

**Comment by ComtelJeremy** [comment]
@thinkingserious Any progress on this? We would like to use it in our app, since we are using SendGrid with an Asp.Net Core web app. Or is there a different solution in the works to use HttpClientFactory?

**Comment by hamakiri** [comment]
Will this be merged at some point ?
Really would like to use that feature without creating my own SendGridClient

**Comment by alex-zwingli** [comment]
@thinkingserious Any updates on this PR?

**Comment by akunzai** [comment]
Hi @thinkingserious,

some conflicts was solved, 

And  the latest commit will automated deploy the `SendGrid.Extensions.DependencyInjection.*.nupkg`.

Are you prefer `SendGrid.Extensions.DependencyInjection` version number begin from 1.0.0 or it should same as SendGrid's version number?

File: ExampleCoreProject/Program.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using SendGrid;
+using SendGrid.Extensions.DependencyInjection;
+using SendGrid.Helpers.Mail;
+
+namespace Example
+{
+    internal class Program
```
Comment (eshanholtz) [inline ExampleCoreProject/Program.cs:11]: Why the rename?

File: .env_sample [hunk 8]
```diff
@@ -1 +1 @@
-SENDGRID_API_KEY=''
+SENDGRID_APIKEY=''
```
Comment (eshanholtz) [inline .env_sample:1]: Revert this env var name change. This is the standard `SENDGRID_API_KEY` env var used in multiple sendgrid helper languages.

File: README.md [hunk 9]
```diff
@@ -8,8 +8,7 @@
 [![GitHub contributors](https://img.shields.io/github/contributors/sendgrid/sendgrid-csharp.svg)](https://github.com/sendgrid/sendgrid-csharp/graphs/contributors)
 [![Open Source Helpers](https://www.codetriage.com/sendgrid/sendgrid-csharp/badges/users.svg)](https://www.codetriage.com/sendgrid/sendgrid-csharp)
 
-<a name="announcements"></a>
-# Announcements
```
Comment (eshanholtz) [inline README.md:12]: nit: put these back, unnecessary change.

File: README.md [hunk 10]
```diff
@@ -61,73 +58,72 @@ Grab your API Key from the [Twilio SendGrid UI](https://app.sendgrid.com/setting
...
 
 The following is the minimum needed code to send an simple email. Use this example, and modify the `apiKey`, `from` and `to` variables:
 
 ```csharp
-using SendGrid;
-using SendGrid.Helpers.Mail;
 using System;
 using System.Threading.Tasks;
+using SendGrid;
+using SendGrid.Helpers.Mail;
```
Comment (eshanholtz) [inline README.md:112]: nit: alphabetize

File: README.md [hunk 11]
```diff
@@ -173,34 +166,76 @@ using System;
...
+        queryParams: queryParams).ConfigureAwait(false);
     }
 }
 ```
+
 ## Web Proxy
 
 ```csharp
-var apiKey = Environment.GetEnvironmentVariable("NAME_OF_THE_ENVIRONMENT_VARIABLE_FOR_YOUR_SENDGRID_KEY");
+var apiKey = Environment.GetEnvironmentVariable("SENDGRID_APIKEY");
```
Comment (eshanholtz) [inline README.md:185]: change to `SENDGRID_API_KEY`

File: ExampleCoreProject/Program.cs (hunk 7 above)
Comment (aevitas) [reply ExampleCoreProject/Program.cs:11]: Am I missing something here? This looks like new code?

File: README.md (hunk 10 above)
Comment (aevitas) [reply README.md:112]: It's fairly standard C# convention to put any custom namespaces after the BCL ones (`System.Foo`), this follows the conventions.

**Comment by aevitas** [comment]
Someone ship this man a medal after this gets merged.

File: README.md (hunk 10 above)
Comment (eshanholtz) [reply README.md:112]: Thank you for the clarification!

File: ExampleCoreProject/Program.cs (hunk 7 above)
Comment (eshanholtz) [reply ExampleCoreProject/Program.cs:11]: It appears this is meant to replace `ExampleCoreProject/Example.cs`?

Comment (akunzai) [reply ExampleCoreProject/Program.cs:11]: @eddiezane in C#, Program.cs is the standard naming for entry point

File: .env_sample (hunk 8 above)
Comment (akunzai) [reply .env_sample:1]: Hi @eshanholtz,

please take a look Samples projects

If we use `SENDGRID_API_KEY` as environment configuration key, in [.NET Platform configuration API](https://docs.microsoft.com/aspnet/core/fundamentals/configuration/), we need use `Configuration["SendGrid:Api:Key"]` to retrieve the APIKey, and JSON configuration will became following example

```json
{
  "SendGrid": {
    "Api":{
     "Key": "CHANGE_ME"
    }
  }
}
```

It create unnecessary depth for configurations.

File: README.md (hunk 9 above)
Comment (akunzai) [reply README.md:12]: I think this is fine to merge header and anchor into single line.

See https://gist.github.com/asabaylus/3071099#gistcomment-2665317 for reference

File: README.md (hunk 11 above)
Comment (akunzai) [reply README.md:185]: see https://github.com/sendgrid/sendgrid-csharp/pull/922#discussion_r430825298

File: .env_sample (hunk 8 above)
Comment (aevitas) [reply .env_sample:1]: @akunzai Isn't it double underscores that create nesting? i.e. `SendGrid__Api__Key` would create a nested structure like you described, whereas `SendGrid_Api_Key` would simply read from a top-level JSON object? 

I personally think having it nested at one level of depth is more practical for most real-world scenarios, as the `SendGrid` configuration section could then hold all configuration related to SG.

Comment (akunzai) [reply .env_sample:1]: @aevitas is right, double underscores create nesting Environment configuration! my bad.

I will drop the commit that Renaming `SENDGRID_API_KEY`

File: ExampleNet45Project/Program.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using SendGrid;
+using SendGrid.Helpers.Mail;
+
+namespace Example
+{
+    internal class Program
```
Comment (eshanholtz) [inline ExampleNet45Project/Program.cs:10]: Really like the simplification you did with these examples. Will you add back the other 'kitchen sink' examples from `ExampleNet45Project/Example.cs` to this file? They don't need to be in both the ExampleCoreProject and `ExampleNet45Project`, but they should be in one of them. They make more sense here, than amongst the `DependencyInjection` usage example.

Comment (akunzai) [reply ExampleNet45Project/Program.cs:12]: @eshanholtz I put back the 'kitchen sink' examples back to `ExampleNet45Project/Example.cs`, thanks for your recommendations.

**Comment by ComtelJeremy** [comment]
Thank you for this, we will put it to use right away. Just wondering when the nuget for SendGrid.Extensions.DependencyInjection will be published. Thank you, again.

**Comment by eshanholtz** [comment]
Next release is scheduled for 2020-06-10

**Comment by arielmoraes** [comment]
Should't the `InjectableSendGridClient` constructor accept an `IOptionsMonitor` instead of `IOptions`? Asking this because the later one doesn't support reloading.

**Comment by akunzai** [comment]
@arielmoraes The [SendGridClientOptions](https://github.com/sendgrid/sendgrid-csharp/blob/master/src/SendGrid/SendGridClientOptions.cs) seems not need to change frequently, and also this PR was closed. (your should join to review this PR earlier)

If it happens in your scenario, another PR is welcome!
